syntax = "proto3";
package pain.proto.manusya;

import "pain/proto/common.proto";

option cc_generic_services = true;

message ChunkOptions {};

message CreateChunkRequest {
    ChunkOptions chunk_options = 1;
    uint32 partition_id = 2;
};

message CreateChunkResponse {
    Header header = 1;
    ObjectId chunk_id = 2;
};

message AppendChunkRequest {
    ObjectId chunk_id = 1;
    uint64 offset = 2;
    uint32 length = 3;
    uint32 crc32 = 4;
};

message AppendChunkResponse {
    Header header = 1;
    uint64 offset = 2;
};

message QueryAndSealChunkRequest {
    ObjectId chunk_id = 1;
};

message QueryAndSealChunkResponse {
    Header header = 1;
    uint64 size = 2;
};

message RemoveChunkRequest {
    ObjectId chunk_id = 1;
};

message RemoveChunkResponse {
    Header header = 1;
};

message ListChunkRequest {
    ObjectId start = 1;
    uint32 limit = 2;
};

message ListChunkResponse {
    Header header = 1;
    bool has_more = 2;
    repeated ObjectId chunk_ids = 3;
};

message ReadChunkRequest {
    ObjectId chunk_id = 1;
    uint64 offset = 2;
    uint32 length = 3;
};

message ReadChunkResponse {
    Header header = 1;
    uint64 offset = 2;
    uint32 length = 3;
    uint32 crc32 = 4;
};

message QueryChunkRequest {
    ObjectId chunk_id = 1;
};

message QueryChunkResponse {
    Header header = 1;
    ChunkOptions chunk_options = 2;
    uint64 size = 3;
    bool sealed = 4;
};

service ManusyaService {
    rpc CreateChunk(CreateChunkRequest) returns (CreateChunkResponse);
    rpc AppendChunk(AppendChunkRequest) returns (AppendChunkResponse);
    rpc QueryAndSealChunk(QueryAndSealChunkRequest)
        returns (QueryAndSealChunkResponse);
    rpc RemoveChunk(RemoveChunkRequest) returns (RemoveChunkResponse);
    rpc ListChunk(ListChunkRequest) returns (ListChunkResponse);
    rpc ReadChunk(ReadChunkRequest) returns (ReadChunkResponse);
    rpc QueryChunk(QueryChunkRequest) returns (QueryChunkResponse);
};
